File: commons/src/main/java/com/tangula/android/utils/TaskUtils.java
package com.tangula.android.utils;

import android.os.AsyncTask;

public class TaskUtils {

    @SuppressWarnings("unused")
    public static void runInUiThread(final Runnable task) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                if (task != null) {
                    task.run();
                }
            }
        }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    @SuppressWarnings("unused")
    public static void runInBackground(final Runnable task) {
        runInAsyncPool(task);
    }

    @SuppressWarnings("WeakerAccess")
    public static void runInAsyncPool(final Runnable task) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                if (task != null) {
                    task.run();
                }
                return null;
            }
        }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    @SuppressWarnings("unused")
    public static void runInSerialExecutor(final Runnable task) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                if (task != null) {
                    task.run();
                }
                return null;
            }
        }.executeOnExecutor(AsyncTask.SERIAL_EXECUTOR);
    }

}

File: commons/src/main/java/com/tangula/android/utils/TaskUtils.kt
package com.tangula.android.utils

import android.annotation.SuppressLint
import android.os.AsyncTask

@Suppress("UNUSED", "MemberVisibilityCanBePrivate")
class TaskUtils {

    companion object {

        /**
         * To start a thread in UI thread.
         * <p>This function use AsyncTask to run the task.</p>
         * @param task The task which would be run.
         */
        @JvmStatic
        @SuppressLint("StaticFieldLeak")
        fun runInUiThread(task:Runnable?){
            object: AsyncTask<Unit, Unit, Unit>(){
                override fun doInBackground(vararg params: Unit?){
                }
                override fun onPostExecute(result: Unit?) {
                    super.onPostExecute(result)
                    task?.run()
                }
            }.execute()
        }

        /**
         * To start a thread in UI thread.
         * <p>This function use AsyncTask to run the task.</p>
         * @param task The task which would be run.
         */
        @JvmStatic
        fun runInUiThread(task:(()->Unit)?){
            runInUiThread(Runnable{task?.invoke()})
        }

        /**
         * To start a thread in background.
         * <p>This function use AsyncTask to run the task.</p>
         * @param task The task which would be run.
         */
        @JvmStatic
        @SuppressLint("StaticFieldLeak")
        fun runInBackground(task: ()->Unit){
            object: AsyncTask<Unit, Unit, Unit>(){
                override fun doInBackground(vararg params: Unit?) {
                    task()
                }
            }.execute()
        }

        /**
         * To start a thread in background.
         * <p>This function use AsyncTask to run the task.</p>
         * @param task The task which would be run.
         */
        @JvmStatic
        fun runInBackground(task: Runnable?){
            runInBackground{task?.run()}
        }

    }
}